### PR TITLE
Test device ID and number of threads after initialization

### PR DESCRIPTION
**Pull request by dalg24**
Intended as a replacement for 
https://github.com/kokkos/kokkos/blob/839df99bcfc94fce4055aa629048b3a7f5b61eee/core/unit_test/CMakeLists.txt#L881-L888

### CODE REVIEW

File: core/unit_test/CMakeLists.txt [hunk 1]
```diff
@@ -955,6 +955,21 @@ KOKKOS_ADD_EXECUTABLE_AND_TEST(
...
 )
 
+add_executable(KokkosCore_UnitTest_DeviceAndThreads UnitTest_DeviceAndThreads.cpp)
+target_link_libraries(KokkosCore_UnitTest_DeviceAndThreads Kokkos::kokkoscore)
+find_package(Python3 COMPONENTS Interpreter)
+if(Python3_Interpreter_FOUND)
+  file(GENERATE
+    OUTPUT $<TARGET_FILE_DIR:KokkosCore_UnitTest_DeviceAndThreads>/TestDeviceAndThreads.py
+    INPUT TestDeviceAndThreads.py
+    USE_SOURCE_PERMISSIONS
```
Comment (dalg24) [inline core/unit_test/CMakeLists.txt:965]: This was added in CMake 3.20

**Review by Rombur — CHANGES_REQUESTED** [review]

File: core/unit_test/UnitTest_DeviceAndThreads.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
+#include <Kokkos_Core.hpp>
```
[the remark this answers is not in the document]
Comment (JBludau) [reply core/unit_test/UnitTest_DeviceAndThreads.cpp:45]: done in [6d4b49](https://github.com/kokkos/kokkos/pull/5328/commits/6d4b49b1bd7421fa12f7d65c36c994c8140772c5)

File: core/unit_test/TestDeviceAndThreads.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import unittest
```
[the remark this answers is not in the document]
Comment (JBludau) [reply core/unit_test/TestDeviceAndThreads.py:43]: done in [21807a](https://github.com/kokkos/kokkos/pull/5328/commits/21807ab584a394a3790aae19b9034e44cc443218)

File: core/unit_test/TestDeviceAndThreads.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        raise Exception(p.stderr.decode("utf-8"))
+    return int(p.stdout)
+
+
+class KokkosInitializationTestCase(unittest.TestCase):
+    def test_num_threads(self):
+        max_threads = GetFlag("max_threads")
+        if max_threads == 1:
+            self.skipTest("no host parallel backend enabled")
+        for num_threads in range(1, max_threads + 1):
```
Comment (crtrott) [inline core/unit_test/TestDeviceAndThreads.py:63]: Let not do every single one. just choose like 3 or so, say 1, 3 and max_threads (3 so we have a non-power of two)

**Review by crtrott — CHANGES_REQUESTED** [review]

**Review by Rombur — APPROVED** [review]

**Review by JBludau — APPROVED** [review]

**Comment by dalg24** [comment]
Didn't mean to re-request Bruno's review...
Christian please go ahead and suggest changes on the PR if you really want to reduce further the number of tests that set num_threads

**Pull request merged**